File: src/main/java/leetcode/problems/LeetCode21.java
package leetcode.problems;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/10/11 22:28
 */
public class LeetCode21 {


    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode();
        mergeTwoLists(list1, list2, dummy);
        return dummy.next;
    }

    public void mergeTwoLists(ListNode list1, ListNode list2, ListNode cur) {
        if (list1 == null) {
            cur.next = list2;
            return;
        }
        if (list2 == null) {
            cur.next = list1;
            return;
        }
        if (list1.val < list2.val) {
            cur.next = list1;
            mergeTwoLists(list1.next, list2, cur.next);
        } else {
            cur.next = list2;
            mergeTwoLists(list1, list2.next, cur.next);
        }
    }
}

File: go/src/leetcode/hot100/LeetCode21.go
package hot100

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	var dfs func(list1, list2, node *ListNode)
	dfs = func(list1, list2, node *ListNode) {
		if list1 == nil {
			node.Next = list2
			return
		}
		if list2 == nil {
			node.Next = list1
			return
		}

		if list1 == nil && list2 == nil {
			return
		}
		if list1.Val < list2.Val {
			node.Next = list1
			dfs(list1.Next, list2, node.Next)
		} else {
			node.Next = list2
			dfs(list1, list2.Next, node.Next)
		}
	}
	dfs(list1, list2, dummy)
	return dummy.Next
}
